metadataserver/cni/env.go

```diff
@@ -0,0 +1,249 @@
+		}
+
+		fd, err := os.OpenFile(fname, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0644) // nolint: gosec
+		if err != nil {
+			return "", err
+		}
+
+		if _, err = fmt.Fprint(fd, uuid.New().String()); err != nil {
+			return "", err
+		}
+
+		if err = fd.Close(); err != nil {
+			return "", err
+		}
+	}
+
+	data, err := ioutil.ReadFile(fname)
+	return string(data), err
+}
+
+func extractEnv(prev *current.Result, pod *v1.Pod) (map[string]string, error) {
+	env := map[string]string{}
+
+	env["LISTEN_PORT"] = "80"
+
+	env["PEER_NAMESPACE"] = getPeerNsPath(pod.Name)
+	env["EC2_REGION"] = os.Getenv("EC2_REGION")
+
+	var ok bool
+
+	for _, ip := range prev.IPs {
+		switch ip.Version {
+		case "4":
+			env["EC2_LOCAL_IPV4"] = ip.Address.String()
+			ok = true
+		case "6":
+			env["EC2_IPV6S"] = ip.Address.String()
+			ok = true
+		}
+	}
+
+	if !ok {
+		return nil, errors.New("No IP information in chained CNI result")
+	}
+
+	// pod specific configuration
+	env["TITUS_TASK_INSTANCE_ID"] = pod.Name
+
+	// FIXME(manas) make this an annotation in the admission webhook
+	env["REQUIRE_TOKEN"] = pod.Annotations[mt.RequireTokenAnnotation]
+	salt, err := generateTokenKeySaltIfRequired(pod)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to generate required token salt")
+	}
+	env["TOKEN_KEY_SALT"] = salt
+
+	env["TITUS_IAM_ROLE"] = pod.Annotations[mt.IamRoleArnAnnotation]
+	env["X_FORWARDED_FOR_BLOCKING_MODE"] = pod.Annotations[mt.XForwardedForBlockingModeAnnotation]
+
+	// assume we are on a metatron enabled host if we find /metatron
+	_, err = os.Stat("/metatron")
+	if os.IsNotExist(err) {
+		env[mt.TitusMetatronVariableName] = "false"
+	}
+	env[mt.TitusMetatronVariableName] = "true"
+
+	return env, nil
+}
+
+func setupEnv(pod *v1.Pod, env map[string]string) error {
+	err := writeEnvFile(pod, env)
+	if err != nil {
+		return errors.Wrap(err, "Unable to write environment file")
+	}
+
+	metatronEnv := env[mt.TitusMetatronVariableName]
+
+	if len(metatronEnv) == 0 {
+		return nil
+	}
+
+	metatronOn, err := strconv.ParseBool(metatronEnv)
+	if err != nil {
+		return errors.Wrap(err, "Unable to tell if metatron is enabled")
+	}
+
+	if !metatronOn {
+		return nil
+	}
+
+	err = writeContainerInfo(pod)
+	if err != nil {
+		return errors.Wrap(err, "Unable to write containerInfo file")
+	}
+
+	return nil
+}
+
+func writeEnvFile(pod *v1.Pod, env map[string]string) error {
+	dname := path.Join(tt.TitusEnvironmentsDir, pod.Name)
+	fname := path.Join(dname, "imds-proxy.env")
+
+	err := os.Mkdir(dname, 0644)
+	if err != nil && !os.IsExist(err) {
+		return err
+	}
+
+	fd, err := os.OpenFile(fname, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644) // nolint: gosec
+	if err != nil {
+		return err
+	}
+
+	for key, val := range env {
+		qval := strconv.QuoteToASCII(val)
+
+		_, err := fmt.Fprintf(fd, "%s=%s\n", key, qval)
+		if err != nil {
+			return err
+		}
+	}
+	return fd.Close()
+}
+
+func writeContainerInfo(pod *v1.Pod) error {
+	str, ok := pod.Annotations["containerInfo"]
```

Comment:
I think this should definitely be a CRI wrapper or OCI hook in creation of the pod sandbox

Reply:
that, or everything that relies on this info should just go back to the kubelet API.